x/epochstorage/keeper/stake_entries.go

```diff
@@ -0,0 +1,192 @@
+package keeper
+
+import (
+	"strconv"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/epochstorage/types"
+)
+
+/* ########## StakeEntry ############ */
+
+// GetStakeEntry gets a specific stake entry from the stake entries KV store
+// Since the stake entries KV store's key includes the provider's stake (which is not known), we iterate over all
+// the providers with the same epoch and chainID and compare the requested address to find the provider
+func (k Keeper) GetStakeEntry(ctx sdk.Context, epoch uint64, chainID string, provider string) (types.StakeEntry, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.StakeEntriesPrefix)
+	iterator := sdk.KVStorePrefixIterator(store, types.StakeEntryKeyPrefixEpochChainId(epoch, chainID))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.StakeEntry
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		if val.Address == provider {
+			return val, true
+		}
+	}
+
+	return types.StakeEntry{}, false
+}
+
+// Set stake entry
+func (k Keeper) SetStakeEntry(ctx sdk.Context, epoch uint64, stakeEntry types.StakeEntry) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.StakeEntriesPrefix)
+	b := k.cdc.MustMarshal(&stakeEntry)
+	store.Set(types.StakeEntryKey(epoch, stakeEntry.Chain, stakeEntry.EffectiveStake(), stakeEntry.Address), b)
+}
+
+// RemoveAllStakeEntriesForEpoch removes all the stake entries of a specific epoch
+func (k Keeper) RemoveAllStakeEntriesForEpoch(ctx sdk.Context, epoch uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.StakeEntriesPrefix)
+	iterator := sdk.KVStorePrefixIterator(store, utils.Serialize(epoch))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+}
+
+// GetAllStakeEntries gets all the stake entries of a specific epoch
+func (k Keeper) GetAllStakeEntriesForGenesis(ctx sdk.Context) []types.StakeStorage {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.StakeEntriesPrefix)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	storagesMap := map[uint64]types.StakeStorage{}
+	for ; iterator.Valid(); iterator.Next() {
+		var entry types.StakeEntry
+		k.cdc.MustUnmarshal(iterator.Value(), &entry)
+		epoch, err := types.ExtractEpochFromStakeEntryKey(string(iterator.Key()))
+		if err != nil {
+			panic(err)
+		}
+		if _, ok := storagesMap[epoch]; !ok {
+			storagesMap[epoch] = types.StakeStorage{
+				Index:          strconv.FormatUint(epoch, 10),
+				EpochBlockHash: k.GetEpochHash(ctx, epoch),
+			}
+		}
+		storage := storagesMap[epoch]
+		storage.StakeEntries = append(storage.StakeEntries, entry)
+		storagesMap[epoch] = storage
+	}
+
+	var storages []types.StakeStorage
+	for _, storage := range storagesMap {
```

Comment:
can you iterate over the keys and sort, in case someone decides to use this in state

Reply:
done. see https://github.com/lavanet/lava/commit/b5ca12d6c32584a95cad6b4e30e1f91cc610b0bf